cumulus/pallets/parachain-system/src/validate_block/trie_recorder.rs

```diff
@@ -0,0 +1,284 @@
+// along with Cumulus. If not, see <http://www.gnu.org/licenses/>.
+
+use codec::Encode;
+
+use sp_std::{
+	cell::{RefCell, RefMut},
+	collections::{btree_map::BTreeMap, btree_set::BTreeSet},
+	rc::Rc,
+};
+use sp_trie::{NodeCodec, ProofSizeProvider, StorageProof};
+use trie_db::{Hasher, RecordedForKey, TrieAccess};
+
+/// A trie recorder that only keeps track of the proof size.
+///
+/// The internal size counting logic should align
+/// with ['sp_trie::recorder::Recorder'].
+pub(crate) struct SizeOnlyRecorder<'a, H: Hasher> {
+	seen_nodes: RefMut<'a, BTreeSet<H::Out>>,
+	encoded_size: RefMut<'a, usize>,
+	recorded_keys: RefMut<'a, BTreeMap<Rc<[u8]>, RecordedForKey>>,
+}
+
+impl<'a, H: trie_db::Hasher> trie_db::TrieRecorder<H::Out> for SizeOnlyRecorder<'a, H> {
+	fn record(&mut self, access: TrieAccess<'_, H::Out>) {
+		let mut encoded_size_update = 0;
+		match access {
+			TrieAccess::NodeOwned { hash, node_owned } =>
+				if !self.seen_nodes.get(&hash).is_some() {
+					let node = node_owned.to_encoded::<NodeCodec<H>>();
+					encoded_size_update += node.encoded_size();
+					self.seen_nodes.insert(hash);
+				},
+			TrieAccess::EncodedNode { hash, encoded_node } => {
+				if !self.seen_nodes.get(&hash).is_some() {
+					let node = encoded_node.into_owned();
+
+					encoded_size_update += node.encoded_size();
+					self.seen_nodes.insert(hash);
+				}
+			},
+			TrieAccess::Value { hash, value, full_key } => {
+				if !self.seen_nodes.get(&hash).is_some() {
+					let value = value.into_owned();
+					encoded_size_update += value.encoded_size();
+					self.seen_nodes.insert(hash);
+				}
+				self.recorded_keys
+					.entry(full_key.into())
+					.and_modify(|e| *e = RecordedForKey::Value)
+					.or_insert_with(|| RecordedForKey::Value);
+			},
+			TrieAccess::Hash { full_key } => {
+				self.recorded_keys
+					.entry(full_key.into())
+					.or_insert_with(|| RecordedForKey::Hash);
+			},
+			TrieAccess::NonExisting { full_key } => {
+				self.recorded_keys
+					.entry(full_key.into())
+					.and_modify(|e| *e = RecordedForKey::Value)
+					.or_insert_with(|| RecordedForKey::Value);
+			},
+		};
+
+		*self.encoded_size += encoded_size_update;
+	}
+
+	fn trie_nodes_recorded_for_key(&self, key: &[u8]) -> RecordedForKey {
+		self.recorded_keys.get(key).copied().unwrap_or(RecordedForKey::None)
+	}
+}
+
+#[derive(Clone)]
+pub(crate) struct SizeOnlyRecorderProvider<H: Hasher> {
+	seen_nodes: Rc<RefCell<BTreeSet<H::Out>>>,
+	encoded_size: Rc<RefCell<usize>>,
+	recorded_keys: Rc<RefCell<BTreeMap<Rc<[u8]>, RecordedForKey>>>,
+}
+
+impl<H: Hasher> SizeOnlyRecorderProvider<H> {
+	pub fn new() -> Self {
+		Self {
+			seen_nodes: Default::default(),
+			encoded_size: Default::default(),
+			recorded_keys: Default::default(),
+		}
+	}
+}
+
+impl<H: trie_db::Hasher> sp_trie::TrieRecorderProvider<H> for SizeOnlyRecorderProvider<H> {
+	type Recorder<'a> = SizeOnlyRecorder<'a, H> where H: 'a;
+
+	fn drain_storage_proof(self) -> StorageProof {
```

Comment:
Dumb question. Would be too much revolutionary to return an `Option<StorageProof>` here?

I'd fit nicely [here](https://github.com/paritytech/polkadot-sdk/blob/372929fa256f5839047bb5d569ad523cb4b15094/substrate/primitives/state-machine/src/trie_backend.rs#L362)